File: go/Generate_Parentheses.go
package main

func generateParenthesis(n int) []string {
	var ans []string
	back(n, n, "", &ans)
	return ans
}

func back(l, r int, cur string, ans *[]string) {
	if l == 0 && r == 0 {
		*ans = append(*ans, cur)
	}

	if l == 0 {
		for r > 0 {
			cur += ")"
			r--
		}
		*ans = append(*ans, cur)
		return
	}

	if r == 0 {
		return
	}

	back(l-1, r, cur+"(", ans)
	if l == r {
		return
	}

	back(l, r-1, cur+")", ans)
}

File: cpp/Generate_Parentheses.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  vector<string> generateParenthesis(int n) {
    vector<string> ans;
    recur(ans, "", n, n);
    return ans;
  }

  void recur(vector<string> &v, string cur, int l, int r) {
    if (l == 0) {
      for (int i = 0; i < r; i++) {
        cur += ")";
      }
      v.push_back(cur);
      return;
    }
    recur(v, cur + "(", l - 1, r);
    if (l < r) {
      recur(v, cur + ")", l, r - 1);
    }
  }
};
